### DISCUSSION TITLE
Calculating magnetization for a transverse Ising Chain

### DISCUSSION THREAD

**Discussion by ArnabeshS** [QuSpin/QuSpin#533] [General]
The Hamiltonian in Spin basis for a transverse Ising chain is given by the Hamiltonian with sigma_z-sigma_z nearest-neighbour interaction and a transverse term sigma_x. The magnetization can be calculated with sigma_z being the order parameter. My question is how this can be implemented in QuSpin using inbuilt classes viz. quantum_operator? Is this can be calculated, more complicated order parameters can be calculated.

**Comment by JiahaoYao** [comment]
> The Hamiltonian in Spin basis for a transverse Ising chain is given by the Hamiltonian with sigma_z-sigma_z nearest-neighbour interaction and a transverse term sigma_x. The magnetization can be calculated with sigma_z being the order parameter. My question is how this can be implemented in QuSpin using inbuilt classes viz. quantum_operator? Is this can be calculated, more complicated order parameters can be calculated. 
> 
> 

Could u give some concrete mathematical formula of the Hamiltonian?

**Reply by ArnabeshS** [reply]
![transverse ISing model](https://user-images.githubusercontent.com/34402914/168528769-f6067a3d-4f55-44f2-9176-733cf019c117.png)
Consider this with J^y=0

**Comment by JiahaoYao** [comment]
Thanks @ArnabeshS , and `Quspin`'s benefit is to make this quite easy and almost just directly translate this to python.

**Comment by JiahaoYao** [comment]
first, you define the coefficient on the site: 
```h_x=[[hx,i] for i in range(L)]```
Then you combine the coefficients with the operator and add to the static list `["x",h_x]`.

**Reply by ArnabeshS** [reply]
actually, for the order parameter it will be `h_z=[[1.0,i] for i in range(L)]` and then the static list `["z",h_z]` using which we can build the quantum_operator ('O' say). If however, we take the expectation, then however things don't quite match up. with increasing H_x, the expectation should fall from one to zero exponentially. This does not quite happen, if the inbuilt methods of quantum_operator class are used.

**Comment by JiahaoYao** [comment]
interesting, it would be good to post your code, and the results, which do not match your thought? (here or issue)? maybe @mgbukov will take a look then.

**Reply by JiahaoYao** [reply]
you might also find this example (https://weinbe58.github.io/QuSpin/examples/example0.html#example0-label) helpful!

**Reply by ArnabeshS** [reply]
Here is a code that I came up with @mgbukov . You will notice that for taking the expectation of sigma_z with respect to the ground state, I had to write two other functions. My question is, is there a way to utilize the inbuilt classes of QuSpin to do the same thing? I also tried using the `quantum_operator` class and its `expt_value` function, but the result is not the expected physically correct result. Thank you for looking into the problem. Let me know if anything is required. I am also attaching the plot which comes as an output:

Code:
```
# Code to calculate magnetization with changing transverse term
from quspin.operators import hamiltonian # Hamiltonians and operators
from quspin.basis import spin_basis_1d # Hilbert space spin basis
import numpy as np # generic math function
import matplotlib.pyplot as plt


def basis2spin(basis,L):
    # method to return a boolean array after converting the passed basis number to binary
    # L is the number os spins in the system
    # basis is state passed
    bin_number=np.binary_repr(basis, width=L)
    a=np.ones(L,dtype=bool)
    k=0
    for i in bin_number:
        if i=='1':
            a[k]=True
        else:
            a[k]=False
        k=k+1
    return a
# method to calculate magnetization(sigma_z)
def magnetization(bstate,state):
    # state has the probabilities after mapping the initial state to the transformed state the mapping being in the same space V
    # bstate has the spins, false is down, true up
    
    
    M=0
    
    #l=state.size
    for i in range(len(bstate)):
        bstate_M=0
        for j in range(bstate[i].size):
            ''' here as sigma_z only gives the sign of the state,
            hence this line is written thus to do the dot product accordingly
            for operators of other kind viz. X and Y care must also be taken for the]
            change of state and then doing the dot product accordingly such that 
            we get the off-diagonal terms. '''
            # bstate_M += (inner product)*(effect of operators)
            bstate_M += ((state[i]**2)*(1 if bstate[i][j] else -1))
                         
                         
        M += abs(bstate_M)
        
    return M

#Here L is the system size, defining a number of spins present, with increasing the system size, a sharp change can be seen.
for L in [8,12]:
    basis=spin_basis_1d(L,pauli=True)
    
    J=-1.0
    h=np.linspace(0.0,4.0,15)
    y=[]
    for lda in h:
        #field=[[g,i] for i in range(L-1)]
        field_zz=[[J,i,(i+1)%L] for i in range(L)]
        #static=[["I",field]]
        field_x=[[lda*(-1.0),i] for i in range(L)]
        static=[["zz",field_zz],["x",field_x]]
        dynamic=[]
        H=hamiltonian(static,dynamic,basis=basis,dtype=np.float64)
        #print(h)
        E,V=H.eigh()
        
        
        #print(E)
        #print(V)
        bstate=[]   # bstate is a list of numpy array
        for i in range(np.power(2,L)):
            bstate.append(basis2spin(basis[i],L))
        
        
        M=magnetization(bstate,V[:,0])/(L*1.0)
        y.append(M)
    
    plt.plot(h,y,label=str(L))
   
plt.xscale('log')
plt.xlabel('Transverse term')
plt.ylabel('Magnetization')
plt.title('$J=-1.0$')
plt.legend()
plt.show()
plt.close()
```

Output:
![plot](https://user-images.githubusercontent.com/34402914/168570634-9d0292ad-408d-48dc-b796-311f5f1244b5.png)

**Reply by ArnabeshS** [reply]
Here is my code by trying the `expt_value` function and the associated plot. The result is not correct physically

```from quspin.operators import hamiltonian,quantum_operator # Hamiltonians and operators
from quspin.basis import spin_basis_1d # Hilbert space spin basis
import numpy as np # generic math function
import os
import matplotlib.pyplot as plt
os.environ['OMP_NUM_THREADS']='12' # set number of OpenMP threads to run in parallel

L=12
basis=spin_basis_1d(L,pauli=True)

J_zz=-1.0
h=np.linspace(0.0,1.0,11) # parameter
y=[]
for h_x in h:
    
    field_zz=[[J_zz,i,(i+1)%L] for i in range(L)]
    field_x=[[h_x,i] for i in range(L)]
    static=[["zz",field_zz],["x",field_x]]
    dynamic=[]
    H=hamiltonian(static,dynamic,basis=basis,dtype=np.float64)
    E,V=H.eigh()
    J_z=[[1.0,i] for i in range(L)]
    static=[["z",J_z]]
    operator_dict_1=dict(H0=static)
    ###### setting up `quantum_operator` hamiltonian dictionary
    O=quantum_operator(operator_dict_1,basis=basis)
    B=O.expt_value(V[:,0])/L
    y.append(B.real)
    
plt.plot(h,y)
plt.xlabel('transverse term')
plt.ylabel('magnetization')
plt.savefig('plot.png')
plt.close()
```

![plot](https://user-images.githubusercontent.com/34402914/168607595-982808ec-d32c-4bdd-8e41-61dd608644ed.png)

**Comment by weinbe58** [comment]
First, the results you show are perfectly reasonable, the problem you are having is that in one case you're plotting |m| while in the other case you are plotting m.  For finite hx the ground state in a finite-size system always obeys the Z2 symmetry sigma^z -> - sigma^z  due to an exponentially small gap between the symmetric and anti-symmetric combination of the two states which have a definite magnetization value pointing in the opposite directions. In the case where hx=0 the gap between these states is exactly 0 leading to a degenerate manifold. 

What you find when doing exact diagonalization on small system is that the symmetry appears to be "broken" for some finite value of hx. This is simply because at a certain point with finite precision arithmetic the gap is too small to be recognized as being finite, in which case, the diagonalization code will choose some random way of orthogonalizing what it believes to be a degenerate subspace and so the eigenvectors will not necessarily obey the Z2 symmetry. I will demonstrate in the code below below. 

Secondly, it not possible to calculate |m_z| because it does not simple function of the pauli operators. You can calculate though m_z^2 which can be expressed as quadratic in sigma^z. For most application this is sufficient, however you can still calculate the operator |m_z| by explicitly constructing the operator m_z. Because m_z is diagonal in the S^z basis we can simply calculate the absoluate value of the diagonal part of the matrix and put it back into a new diagonal matrix. At that point you can pass this new matrix back into the QuSpin classes by replacing the operator strings with the explicit matrix you have constructed. In the code below I do this exactly. 

```python

from quspin.basis import spin_basis_general
from quspin.operators import quantum_operator,hamiltonian
from scipy.sparse import dia_matrix
from scipy.sparse.linalg import eigsh
import numpy as np
import matplotlib.pyplot as plt



def get_m(basis):
    h_list = [[1.0/L,i] for i in range(L)]
    kwargs=dict(basis=basis,dtype=np.float64,
        check_symm=False,check_pcon=False,
        check_herm=False)

    m = hamiltonian([["z",h_list]],[],**kwargs)

    shape = (basis.Ns,basis.Ns)

    abs_m = dia_matrix((np.abs(m.diagonal()).reshape((1,-1)),[0]),shape=shape,dtype=m.dtype)
    abs_m = hamiltonian([abs_m],[],**kwargs)

    return m,abs_m

L = 12
basis = spin_basis_general(L,pauli=True)

h_list = [[1.0,i] for i in range(L)]
J_list = [[1.0,i,i+1] for i in range(L-1)]

op_dict = dict(J=[["zz",J_list]],hx=[["x",h_list]],hz=[["z",h_list]])

kwargs=dict(basis=basis,dtype=np.float64,
    check_symm=False,check_pcon=False,
    check_herm=False)

H = quantum_operator(op_dict,**kwargs)
M,abs_M = get_m(basis)

# calculating m^2
m2_list = [[1.0/L**2,i,j] for i in range(L) for j in range(L)]
M2 = hamiltonian([["zz",m2_list]],[],**kwargs)

hxs = np.linspace(0,2,101)

m2 = []
abs_m = []
m = []
gap = []

f,(ax1,ax2) = plt.subplots(2,1,sharex=True)

for hx in hxs:

	e,v = eigsh(H.aslinearoperator(pars=dict(J=-1,hx=hx,hz=0)),k=2,which="SA")

	psi0 = v[:,0]
	gap.append(np.abs(e[1]-e[0]))

	abs_m.append(abs_M.expt_value(psi0))
	m.append(M.expt_value(psi0))
	m2.append(M2.expt_value(psi0))

ax1.plot(hxs,abs_m,label="$|m|$")
ax1.plot(hxs,m,label="$m$")
ax1.plot(hxs,m2,label="$m^2$")
ax1.legend()
ax2.plot(hxs,gap)
ax2.set_ylabel("gap")
ax2.set_yscale("log")
f.tight_layout()
f.subplots_adjust(hspace=0.02)
plt.show()

```

The resulting plot is given by:

![tfim](https://user-images.githubusercontent.com/10360423/168626345-6930ae5b-1ae3-42ac-8494-a1f800e6493b.png)

You can see that once the gap gets to be too small, the value of m starts to flucuate wildly due to the orthogonalization of this degenerate subspace. 

FYI: if you want code to show up formatted correctly put it inside follow this syntax: https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code

**Reply by ArnabeshS** [reply]
Thank you @weinbe58 , for your technical and physical insight!
